clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -276,11 +289,16 @@ private static FetchResponseData toMessage(Errors error,
                     .setPartitions(partitionResponses));
             }
         }
-
-        return new FetchResponseData()
-            .setThrottleTimeMs(throttleTimeMs)
-            .setErrorCode(error.code())
-            .setSessionId(sessionId)
-            .setResponses(topicResponseList);
+        data.setThrottleTimeMs(throttleTimeMs)
+                .setErrorCode(error.code())
+                .setSessionId(sessionId)
+                .setResponses(topicResponseList);
+        nodeEndpoints.forEach(endpoint -> data.nodeEndpoints().add(
```

Comment:
I don't think we need to populate nodeEndpoints field for all the Fetch/Produce response, which has certain overhead.  to my understanding, the info needed only for error condition.

Reply:
that should be getting handled in KafkaApis where we only add to nodeEndpoints if there is an error, otherwise it should be an empty list.

Reply:
it makes sense, thanks

Reply:
Nit: It would be good to add a comment here, something like
KafkaApis will manage the response, returning nodeEndpoints information only in case of an error.

Reply:
added

Reply:
Could we also create the nodeEndpoints together so we keep the final data object relatively clean. (Ie we generate the response data structure above, we could also generate the enpoints data structure above.)

Reply:
reordered to make things cleaner